src/librbd/cache/pwl/rwl/WriteLog.cc

```diff
@@ -117,6 +117,10 @@ void WriteLog<I>::alloc_op_log_entries(GenericLogOperations &ops)
     m_log_entries.push_back(log_entry);
     ldout(m_image_ctx.cct, 20) << "operation=[" << *operation << "]" << dendl;
   }
+  if (m_cache_state->empty && m_first_free_entry != m_first_valid_entry) {
```

Comment:
> I used to call `m_log_entries.empty()`, but there was a problem here(`retire_entries`),  `m_log_entries` and `m_first_valid_entry` are not changed in a code block, I can only update it after `m_first_valid_entry` is modified, but then `m_log_entries` may have been change because `m_lock` don't lock two code blocks. 
>
> If the display here does not need to be so strict, it is also ok to use `m_log_entries.empty()`.

If `m_log_entries` becomes empty and then, as soon as the lock is released, another thread adds a new log entry, in the end the cache isn't really empty, is it?  Since the image metadata json blob and `m_first_{valid,free}_entry` pointer updates aren't atomic at the RADOS level either, I think it would be better to use `m_log_entries.empty()`.  It seems slightly more on point from the user's perspective to me and also consistent with using `m_dirty_log_entries.empty()` for `clean` tracking.

Reply:
ok, done